### ISSUE TITLE
Typo3 12 Compatibility

### ISSUE THREAD

**Issue by RudolfGross** [a-r-m-i-n/dce#45]
it would be great to get a typo3 12 compatibility for this extension.
Will there be an update for Typo3 12 in the near future? Asking for my next web project.

**Comment by ingeniumdesign** [comment]
New Infos?

**Comment by Codersunlimited** [comment]
Can you already foresee when there will be a version for TYPO3 v12 for this?

**Comment by a-r-m-i-n** [comment]
Planned for beginning of August. And I decided to integrate a side project, which allows you to define DCEs using YAML configuration files, which can get shipped in extensions.

Thanks for your patience.

**Comment by ste101** [comment]
I made an upgrade and tried to use the v12 version.
Backend module seems to work right.
When I click on page or list I'm getting this error:
`Invoked ContentObjectRenderer::parseFunc without any configuration`

**Comment by a-r-m-i-n** [comment]
@ste101 Do you have a custom Backend template, which uses f:format.html? I did not get this error, yet.

**Comment by a-r-m-i-n** [comment]
@ste101 Yes, with `<f:format.html>...</f:format.html>` in field *Backend template content (fluid)* I am able to reconstruct your issue.

**Comment by a-r-m-i-n** [comment]
DCE 3.0 is out now! 🎉

**Comment by ste101** [comment]
composer.json prevents installing of EXT.
TYPO3 12.4.5
PHP 8.2

**Comment by ste101** [comment]
After installing 'Upgrade Wizard' is broken.

**Comment by a-r-m-i-n** [comment]
@ste101 Please be more specific. What does the error log say?

And did you clear all caches (from install tool)?

**Comment by vincentvega123** [comment]
Good Morning, I can confirm the last two issues here. After downloading dce 3.0.0 from TER I am not able to install the extension (icon is not appearing) unless I delete (or uncomment) the composer.json. After deleting composer.json the plus-icon appears and the installation is possible.

Typo3 12.4.5
PHP 8.1

**Comment by vincentvega123** [comment]
Concerning the Upgrade Wizard I found this in the error log:

Tue, 22 Aug 2023 08:53:59 +0200 [WARNING] request="1c27875780ade" component="TYPO3.CMS.Core.Error.ErrorHandler": Core: Error handler (BE): PHP Warning: file_get_contents(/image-slider.html): Failed to open stream: No such file or directory in /typo3conf/ext/dce/Classes/UpdateWizards/MigrateOldNamespacesInFluidTemplateUpdateWizard.php line 173 
Tue, 22 Aug 2023 08:53:59 +0200 [CRITICAL] request="1c27875780ade" component="TYPO3.CMS.Core.Error.ProductionExceptionHandler": Core: Exception handler (WEB: BE): TypeError, code #0, file /typo3conf/ext/dce/Classes/UpdateWizards/MigrateOldNamespacesInFluidTemplateUpdateWizard.php, line 179: T3\Dce\UpdateWizards\MigrateOldNamespacesInFluidTemplateUpdateWizard::templateNeedUpdate(): Argument #1 ($templateContent) must be of type string, bool given, called in /typo3conf/ext/dce/Classes/UpdateWizards/MigrateOldNamespacesInFluidTemplateUpdateWizard.php on line 173 - {"mode":"WEB","application_mode":"BE","exception_class":"TypeError","exception_code":0,"file":"/typo3conf/ext/dce/Classes/UpdateWizards/MigrateOldNamespacesInFluidTemplateUpdateWizard.php","line":179,"message":"T3\\Dce\\UpdateWizards\\MigrateOldNamespacesInFluidTemplateUpdateWizard::templateNeedUpdate(): Argument #1 ($templateContent) must be of type string, bool given, called in /typo3conf/ext/dce/Classes/UpdateWizards/MigrateOldNamespacesInFluidTemplateUpdateWizard.php on line 173","request_url":"/typo3/install.php?install[controller]=upgrade&install[context]=backend&install[action]=upgradeWizardsList","exception":null}

**Comment by vincentvega123** [comment]
Besides of that, there is no syntax-highlighting in the backend module any more and when I select something from the dropdown configuration. there is nothing added/changed to the textarea below.

![Bildschirmfoto 2023-08-22 um 09 03 48](https://github.com/a-r-m-i-n/dce/assets/10086856/10a09ad6-ee59-4ed9-a9d9-f535e149bb10)

**Comment by vincentvega123** [comment]
Additionally, none of the DCEs is shown in the Wizards for adding content on a page and also not in the dropdown when selecting another content element first. But maybe this is because I would have to execute the Upgrade wizard first?

**Comment by vincentvega123** [comment]
> Besides of that, there is no syntax-highlighting in the backend module any more and when I select something from the dropdown configuration. there is nothing added/changed to the textarea below.
> 
> ![Bildschirmfoto 2023-08-22 um 09 03 48](https://user-images.githubusercontent.com/10086856/262263567-10a09ad6-ee59-4ed9-a9d9-f535e149bb10.png)

Sorry for spaming: I just realize, that changes from the dropdown is working if I add a new DCE and then new fields. Only the existing ones are behaving like described before. I started using the 3.0.0-dev Version before and added DCEs at that time already. That's why they already existed before the final 3.0.0. was released yesterday

**Comment by ste101** [comment]
> @ste101 Please be more specific. What does the error log say?
> 
> And did you clear all caches (from install tool)?

Sorry, I wasn't clear enough.
The problem with the upgrade wizard occures when EXT:container (2.2.5) and EXT:container_elements (5.1.0) is installed.
I either dce oder container_elements is not installed the upgrade wizard works.

Part of error message:
```
(1/1) #1638976490 LogicException
TYPO3\CMS\Core\Database\ConnectionPool can not be injected/instantiated during ext_localconf.php or TCA loading. Use lazy loading instead.

in /homepages/25/d515604501/htdocs/typo3-12/typo3_src-12.4.5/typo3/sysext/core/Classes/ServiceProvider.php line 158

    public static function getConnectionPool(ContainerInterface $container): Database\ConnectionPool
    {
        if (!$container->get('boot.state')->complete) {
            throw new \LogicException(Database\ConnectionPool::class . ' can not be injected/instantiated during ext_localconf.php or TCA loading. Use lazy loading instead.', 1638976490);
        }

        return self::new($container, Database\ConnectionPool::class);
    }
at TYPO3\CMS\Core\ServiceProvider::getConnectionPool()
in /homepages/25/d515604501/htdocs/typo3-12/typo3temp/var/cache/code/di/DependencyInjectionContainer_da53863478884ba9f680509a01cb5f525b840ca9.php line 13453
     * @return \TYPO3\CMS\Core\Database\ConnectionPool
     */
    protected function getConnectionPoolService()
    {
        return $this->services['TYPO3\\CMS\\Core\\Database\\ConnectionPool'] = \TYPO3\CMS\Core\ServiceProvider::getConnectionPool($this);
    }

    /**
     * Gets the public 'TYPO3\CMS\Core\Database\Schema\SqlReader' shared autowired service.
at DependencyInjectionContainer_da53863478884ba9f680509a01cb5f525b840ca9->getConnectionPoolService()
in /homepages/25/d515604501/htdocs/typo3-12/typo3_src-12.4.5/vendor/symfony/dependency-injection/Container.php line 219
        try {
            if (isset($this->fileMap[$id])) {
                return /* self::IGNORE_ON_UNINITIALIZED_REFERENCE */ 4 === $invalidBehavior ? null : $this->load($this->fileMap[$id]);
            } elseif (isset($this->methodMap[$id])) {
                return /* self::IGNORE_ON_UNINITIALIZED_REFERENCE */ 4 === $invalidBehavior ? null : $this->{$this->methodMap[$id]}();
            }
        } catch (\Exception $e) {
            unset($this->services[$id]);

at Symfony\Component\DependencyInjection\Container->make()
in /homepages/25/d515604501/htdocs/typo3-12/typo3_src-12.4.5/vendor/symfony/dependency-injection/Container.php line 199
    public function get(string $id, int $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE): ?object
    {
        return $this->services[$id]
            ?? $this->services[$id = $this->aliases[$id] ?? $id]
            ?? ('service_container' === $id ? $this : ($this->factories[$id] ?? $this->make(...))($id, $invalidBehavior));
    }

    /**
     * Creates a service.
at Symfony\Component\DependencyInjection\Container->get()
in /homepages/25/d515604501/htdocs/typo3-12/typo3_src-12.4.5/typo3/sysext/core/Classes/Utility/GeneralUtility.php line 2965
        // support classes that require dependency injection.
        // We operate on the original class name on purpose, as class overrides
        // are resolved inside the container
        if (self::$container !== null && $constructorArguments === [] && self::$container->has($className)) {
            return self::$container->get($className);
        }

        // Create new instance and call constructor with parameters
        $instance = new $finalClassName(...$constructorArguments);
at TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance()
in /homepages/25/d515604501/htdocs/typo3-12/typo3conf/ext/dce/Classes/Utility/DatabaseUtility.php line 26
     * Get TYPO3s Connection Pool.
     */
    public static function getConnectionPool(): ConnectionPool
    {
        return GeneralUtility::makeInstance(ConnectionPool::class);
    }

    /**
     * Give it a non executed QueryBuilder to fetch result as array
at T3\Dce\Utility\DatabaseUtility::getConnectionPool()
in /homepages/25/d515604501/htdocs/typo3-12/typo3conf/ext/dce/Classes/Utility/DatabaseUtility.php line 54
     */
    public static function adminGetTables(): array
    {
        $tables = [];
        foreach (static::getConnectionPool()->getConnectionNames() as $connectionName) {
            $connection = static::getConnectionPool()->getConnectionByName($connectionName);
            foreach ($connection->createSchemaManager()->listTableNames() as $tableName) {
                $tables[$tableName] = $tableName;
            }
at T3\Dce\Utility\DatabaseUtility::adminGetTables()
in /homepages/25/d515604501/htdocs/typo3-12/typo3conf/ext/dce/Classes/Components/ContentElementGenerator/InputDatabase.php line 40
     * @return array with DCE -> containing tabs -> containing fields
     */
    public function getDces(): array
    {
        $tables = DatabaseUtility::adminGetTables();
        if (!\array_key_exists('tx_dce_domain_model_dce', $tables) ||
            !\array_key_exists('tx_dce_domain_model_dcefield', $tables)
        ) {
            return [];
at T3\Dce\Components\ContentElementGenerator\InputDatabase->getDces()
in /homepages/25/d515604501/htdocs/typo3-12/typo3conf/ext/dce/Configuration/Icons.php line 33
];

/** @var InputDatabase $dceInputDatabase */
$dceInputDatabase = GeneralUtility::makeInstance(InputDatabase::class);
foreach ($dceInputDatabase->getDces() as $dce) {
    if ($dce['hasCustomWizardIcon'] && !empty($dce['wizard_custom_icon'])) {
        $wizardCustomIcon = $dce['wizard_custom_icon'];

        $iconProvider = BitmapIconProvider::class;
```

**Comment by a-r-m-i-n** [comment]
@vincentvega123 Thanks for reporting. The error in the old upgrade wizard, seems to be an actual issue.

You are right, regarding syntax highlighting. I had to reimplemented the code mirror functionality, because the backend requires ES6 Javascript. I want to implement t3editor support, but this is still on my schedule.

**Comment by a-r-m-i-n** [comment]
The issue, that you weren't able to install DCE in TYPO3 legacy mode is fixed in DCE 3.0.2

Also see #76 for further details

**Comment by ste101** [comment]
> @ste101 Yes, with `<f:format.html>...</f:format.html>` in field _Backend template content (fluid)_ I am able to reconstruct your issue.
Did you find a solution for this issue?
Same problem with mask for backend templates.
